### PR TITLE
Subword Tokenizer HuggingFace like API

### PR DESCRIPTION
**Pull request by VibhuJawa**
This PR closes https://github.com/rapidsai/cudf/issues/5868 by adding a new tokenizer API. 

We are seeing speedups even at low batch sizes (10/100) so this should potentially unlock some inference/training use cases for us. 

## Benchmarks:
(Thanks to  @davidwendt  for writing the super fast tokenizer  :boom: )

| Batch Size 	| HuggingFace 	| Rapids Old API 	| Rapids Tokenizer API 	| Tokenizer API Speed up  vs HuggingFace 	| Rapids New API  Speedup 	|
|-	|-	|-	|-	|-	|-	|
| 1 	| 0.000242 	| 0.006890 	| 0.000497 	| 0.487 	| 13.863 	|
| 10 	| 0.002800 	| 0.007030 	| 0.000516 	| 5.426 	| 13.624 	|
| 100 	| 0.016200 	| 0.007140 	| 0.000537 	| 30.168 	| 13.296 	|
| 1000 	| 0.149000 	| 0.007150 	| 0.000517 	| 288.201 	| 13.830 	|

## API Comparision to HuggingFace:

The goal of this PR is to ensure our API matches up HuggingFace as much as possible to help with ease of porting. 

Proposed API in this PR:

```python
from cudf.core.subword_tokenizer import SubwordTokenizer
tokenizer = SubwordTokenizer('bert-base-cased-vocab-hash.txt',do_lower_case=False)

output = tokenizer(str_series,
                   max_num_rows=len(str_series),
                   truncation=True,
                   max_length=seq_len,
                   padding='max_length',
                   add_special_tokens=False,
                   return_tensors='pt')
```           

HuggingFace API:
```python
from transformers import BertTokenizerFast
tokenizer = BertTokenizerFast.from_pretrained('bert-base-cased', do_lower_case=False)

output = tokenizer(input_sentence_ls,
                   truncation=True,
                   max_length=seq_len,
                   padding='max_length',
                   add_special_tokens=False,
                   return_tensors = 'pt')
output_d = {k:v.cuda() for k,v in output.items()} 
```

## TODO:
- [x] Add tests
- [x] Throw appropriate warnings for HuggingFace discrepancies 
- [x] API checks 
- [X] [ Benchmark/Example Notebook ](https://nbviewer.jupyter.org/gist/VibhuJawa/350a8479b10be3591dd9c4d5da3cfc3b)

CC: @raykallen, @BartleyR  (from the cyber team)

CC: @randerzander , @beckernick  (from the workflows team)

### CODE REVIEW

**Review by VibhuJawa — COMMENTED** [review]
Changed DataFiles to allow for testing against HuggingFace.

File: ci/gpu/build.sh [hunk 1]
```diff
@@ -180,6 +180,7 @@ else
 
     gpuci_logger "Installing $CUDF_CONDA_FILE & $KAFKA_CONDA_FILE"
     conda install -c ${CONDA_ARTIFACT_PATH} "$CUDF_CONDA_FILE" "$KAFKA_CONDA_FILE"
+    conda install -c conda-forge tokenizers
```
Comment (VibhuJawa) [inline ci/gpu/build.sh:183]: Added here temporarily, Will add to integration repo once tests pass.

File: ci/gpu/build.sh [hunk 2]
```diff
@@ -180,6 +180,7 @@ else
 
     gpuci_logger "Installing $CUDF_CONDA_FILE & $KAFKA_CONDA_FILE"
     conda install -c ${CONDA_ARTIFACT_PATH} "$CUDF_CONDA_FILE" "$KAFKA_CONDA_FILE"
+    conda install -c conda-forge transformers
```
Comment (VibhuJawa) [inline ci/gpu/build.sh:183]: Installed temporarily for testing.

**Comment by BartleyR** [comment]
@mdemoret-nv for Morpheus

File: ci/gpu/build.sh (hunk 2 above)
Comment (VibhuJawa) [reply ci/gpu/build.sh:183]: @galipremsagar , What will be the recommended way to add such a testing-only dependency?

Comment (galipremsagar) [reply ci/gpu/build.sh:183]: If this is testing only, these are the places we will have to add `transformers` to:

https://github.com/rapidsai/integration/blob/branch-0.19/conda/recipes/rapids-build-env/meta.yaml

https://github.com/rapidsai/cudf/tree/branch-0.20/conda/environments (in all yml files)

and a new entry here in `setup.py` : https://github.com/rapidsai/cudf/blob/025c56ab1c4046f0f3d1d3ec768b1a69755643c8/python/cudf/setup.py#L35-L44

Comment (VibhuJawa) [reply ci/gpu/build.sh:183]: Thanks for the help Prem, will add it at the needed places.

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            truncation = False
+            warning_msg = (
+                "The behaviour currently differs from  "
+                + "HuggingFace as we always return overflowing tokens"
+            )
+            warn(warning_msg)
+            if add_special_tokens:
+                error_msg = (
+                    "Adding special tokens is not supported"
+                    + f"with truncation = {truncation}"
```
Comment (galipremsagar) [inline python/cudf/cudf/core/subword_tokenizer.py:176]:
```suggestion
                    f"Adding special tokens is not supported"
                    f"with truncation = {truncation}. "
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                )
+                recommendation = (
+                    "Custom Cupy kernel can potentially"
+                    + "be used to add it. For reference "
+                    + "see: _bert_add_special_tokens "
```
Comment (galipremsagar) [inline python/cudf/cudf/core/subword_tokenizer.py:181]:
```suggestion
                    "Custom Cupy kernel can potentially"
                    "be used to add it. For reference "
                    "see: _bert_add_special_tokens "
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            raise NotImplementedError(error_msg)
+
+        if max_length <= stride:
+            error_msg = "Stride should be less than max_length"
+            raise ValueError(error_msg)
+
+        if return_tensors not in ["cp", "pt", "tf"]:
+            error_msg = (
+                "Only cupy(cp), pytorch(pt) and tensorflow(tf)"
+                + "tensors are supported"
```
Comment (galipremsagar) [inline python/cudf/cudf/core/subword_tokenizer.py:198]:
```suggestion
                "Only cupy(cp), pytorch(pt) and tensorflow(tf)"
                "tensors are supported"
```

we can remove the `+` sign here just to make it look indented in code, else python concatenates both ways.

File: python/cudf/cudf/_lib/nvtext/subword_tokenize.pyx [hunk 6]
```diff
@@ -9,27 +9,67 @@ from libc.stdint cimport uintptr_t
...
 
+def subword_tokenize_vocab_file(
+    Column strings,
+    object   hash_file,
+    uint32_t max_sequence_length=64,
+    uint32_t stride=48,
+    bool do_lower=True,
+    bool do_truncate=False,
+    uint32_t max_rows_tensor=500
+):
```
Comment (galipremsagar) [inline python/cudf/cudf/_lib/nvtext/subword_tokenize.pyx:69]: Maybe let's add a docstring here what this method does in short.

File: ci/gpu/build.sh (hunk 2 above)
Comment (VibhuJawa) [reply ci/gpu/build.sh:183]: Have added to setup.py with commit :  https://github.com/rapidsai/cudf/pull/7942/commits/6d3e86c664cb0df4d5913ec01d12afd2c76e3b13  and to conda files  with https://github.com/rapidsai/cudf/pull/7942/commits/4679c859e23b51f37de159850a2037fb4b48d335 .  


Will remove the dependency here as soon as integration PR  https://github.com/rapidsai/integration/pull/249 lands .

**Comment by VibhuJawa** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/setup.py [hunk 7]
```diff
@@ -40,6 +40,7 @@
         "hypothesis" "mimesis",
         "pyorc",
         "msgpack",
+        "transformers",
```
Comment (kkraus14) [inline python/cudf/setup.py:43]: I think this needs to be added to the `requirements.txt` file as well

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+from __future__ import annotations
+import cupy as cp
+from warnings import warn
+from cudf._lib.nvtext.subword_tokenize import (
+    subword_tokenize_inmem_hash as cpp_subword_tokenize,
+    Hashed_Vocabulary as cpp_hashed_vocabulary,
+)
+
+
+def _cast_to_appripate_type(ar, cast_type):
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:12]:
```suggestion
def _cast_to_appropriate_type(ar, cast_type):
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+        tokenizer_output["attention_mask"] = cp.asarray(
+            attention_mask
+        ).reshape(-1, max_length)
+        tokenizer_output["metadata"] = cp.asarray(metadata).reshape(-1, 3)
+
+        if add_special_tokens:
+            tokenizer_output = _bert_add_special_tokens(tokenizer_output)
+
+        tokenizer_output = {
+            k: _cast_to_appripate_type(v, return_tensors)
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:230]:
```suggestion
            k: _cast_to_appropriate_type(v, return_tensors)
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+            overlapping tokens.
+            The information about the overlapping tokens is
+            present in the metadata outputed.
+
+        return_tensors (str) default (cp.ndarray objects)
+            'tf': Return TensorFlow tf.constant objects
+            'pt': Return PyTorch torch.Tensor objects
+            'cp': Return cupy cp.ndarray objects
+
+        return_token_type_ids (bool, optional): Only False currently supported
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:115]: I think these need a bit of formatting love. Needs to have a space before and after the colons for formatting reasons.

Also, is there a reason the types are in parenthesis?

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+                error_msg = (
+                    f"Adding special tokens is not supported"
+                    f"with truncation = {truncation}. "
+                )
+                recommendation = (
+                    "Custom Cupy kernel can potentially"
+                    "be used to add it. For reference "
+                    "see: _bert_add_special_tokens "
+                )
+                raise NotImplementedError(error_msg + recommendation)
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:185]: Should this be evaluated and possibly throw before we raise the warning?

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+
+        if padding != "max_length":
+            error_msg = (
+                "Currently padding to the provided max_length is supported"
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:189]: Should this be *not* supported?
```suggestion
                "Currently padding to the provided max_length is not supported"
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+            )
+            raise NotImplementedError(error_msg)
+
+        if max_length <= stride:
+            error_msg = "Stride should be less than max_length"
+            raise ValueError(error_msg)
+
+        if return_tensors not in ["cp", "pt", "tf"]:
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:197]:
```suggestion
        if return_tensors not in {"cp", "pt", "tf"}:
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,279 @@
...
+        )
+
+        tokenizer_output = dict()
+        tokenizer_output["input_ids"] = cp.asarray(input_ids).reshape(
+            -1, max_length
+        )
+        tokenizer_output["attention_mask"] = cp.asarray(
+            attention_mask
+        ).reshape(-1, max_length)
+        tokenizer_output["metadata"] = cp.asarray(metadata).reshape(-1, 3)
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:224]: Should just initialize the dictionary:
```suggestion
        tokenizer_output = {
            "input_ids": cp.asarray(input_ids).reshape(
                -1, max_length
            )
            "attention_mask": cp.asarray(
                attention_mask
            ).reshape(-1, max_length)
            "metadata": cp.asarray(metadata).reshape(-1, 3)
        }
```

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+
+        tokenizer_output = {
+            k: _cast_to_appripate_type(v, return_tensors)
+            for k, v in tokenizer_output.items()
+        }
+
+        return tokenizer_output
+
+
+def _bert_add_special_tokens(token_o):
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:237]: Can you add a comment or brief docstring explaining what this function does?

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 16]
```diff
@@ -0,0 +1,279 @@
...
+    _bert_add_special_tokens_input_ids(token_o["input_ids"], seq_end_col)
+    _bert_add_special_tokens_attention_mask(
+        token_o["attention_mask"], seq_end_col
+    )
+    _bert_add_spedical_tokens_metadata(token_o["metadata"], max_length)
+
+    return token_o
+
+
+def _bert_add_special_tokens_input_ids(input_ids, seq_end_col):
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:253]: Add a comment or brief docstring explaining what this function does.

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+    input_ids[:, 1:-1] = input_ids[:, 0:-2]
+    input_ids[:, 0] = 101
+    # Mark end of sequence [SEP]
+
+    input_ids[
+        cp.arange(0, input_ids.shape[0], dtype=cp.uint32), seq_end_col
+    ] = 102
+
+
+def _bert_add_special_tokens_attention_mask(attention_mask, seq_end_col):
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:264]: Add comment or brief docstring

File: python/cudf/cudf/core/subword_tokenizer.py [hunk 18]
```diff
@@ -0,0 +1,279 @@
...
+    attention_mask[:, 1:-1] = attention_mask[:, 0:-2]
+    # Mark [CLS] token with 1
+    attention_mask[:, 0] = 1
+    # Mark [SEP] token with 1
+    attention_mask[
+        cp.arange(0, attention_mask.shape[0], dtype=cp.uint32), seq_end_col
+    ] = 1
+
+
+def _bert_add_spedical_tokens_metadata(metadata, max_length):
```
Comment (kkraus14) [inline python/cudf/cudf/core/subword_tokenizer.py:275]: Should this be special instead of spedical?

Also, add comment or brief docstring

**Comment by kkraus14** [comment]
Also, is it necessary to change the text files used for testing here? Changing big files like this just blows up the git repository size over time.

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 8 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Resolved with commit   https://github.com/rapidsai/cudf/pull/7942/commits/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 9 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Resolved with commit https://github.com/rapidsai/cudf/pull/7942/commits/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 11 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Switched it around to throw before we raise the warning with commit https://github.com/rapidsai/cudf/pull/7942/commits/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 12 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Nope, it the only value with support.  I changed the error message to below with commit https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3 now to make it clearer. : 

```
Only padding to the provided max_length is currently supported
```

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 13 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Resolved with commit: https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 14 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Resolved with commit : https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 15 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py:238]: Resolved with commit : https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 16 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py:258]: Resolved with commit https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 17 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py:273]: Resolved with commit : https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 18 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Resolved with commit https://github.com/rapidsai/cudf/commit/0018779bde5984931e114a3cd412ce869706a7b3

**Comment by VibhuJawa** [comment]
> Also, is it necessary to change the text files used for testing here? Changing big files like this just blows up the git repository size over time.

We previously had `uncased` vocabulary which was randomly sampled making it difficult to run tests with it. This new vocabulary is a cased one and does meaningful tokenization.  We can probably just download them on the fly if making this is prohibitive.   FWIW, I don't expect this file to change again in the future.

**Review by ajschmidt8 — APPROVED** [review]
approved now that https://github.com/rapidsai/integration/pull/249 is merged.

File: python/cudf/cudf/core/subword_tokenizer.py (hunk 10 above)
Comment (VibhuJawa) [reply python/cudf/cudf/core/subword_tokenizer.py]: Fixed all documentation and formatting issues with commit https://github.com/rapidsai/cudf/pull/7942/commits/26cc440377ac3f6eaadb9597716e89b073490f25

This is how it is rendered locally: 
image

image

image

File: python/cudf/setup.py (hunk 7 above)
Comment (VibhuJawa) [reply python/cudf/setup.py:43]: Added it with commit https://github.com/rapidsai/cudf/pull/7942/commits/23e9fa437f0ac088ad4b22960d3fcb75bfd0709c

File: python/cudf/requirements/cuda-11.0/dev_requirements.txt [hunk 19]
```diff
@@ -36,6 +36,6 @@ sphinx-copybutton
 sphinx-markdown-tables
 sphinx_rtd_theme
 sphinxcontrib-websupport
-typing_extensions
```
Comment (VibhuJawa) [inline python/cudf/requirements/cuda-11.0/dev_requirements.txt:39]: `typing_extensions ` was duplicated twice so removed it

File: python/cudf/requirements/cuda-11.1/dev_requirements.txt [hunk 20]
```diff
@@ -36,6 +36,6 @@ sphinx-copybutton
 sphinx-markdown-tables
 sphinx_rtd_theme
 sphinxcontrib-websupport
-typing_extensions
```
Comment (VibhuJawa) [inline python/cudf/requirements/cuda-11.1/dev_requirements.txt:39]: `typing_extensions ` was duplicated twice so removed it

File: python/cudf/requirements/cuda-11.2/dev_requirements.txt [hunk 21]
```diff
@@ -36,6 +36,6 @@ sphinx-copybutton
 sphinx-markdown-tables
 sphinx_rtd_theme
 sphinxcontrib-websupport
-typing_extensions
```
Comment (VibhuJawa) [inline python/cudf/requirements/cuda-11.2/dev_requirements.txt:39]: `typing_extensions ` was duplicated twice so removed it

**Comment by VibhuJawa** [comment]
@kkraus14,  @galipremsagar. Thanks a ton for your detailed reviews.  

I think all reviews have been addressed, and this PR is ready for another review.

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by galipremsagar** [comment]
rerun tests

**Comment by VibhuJawa** [comment]
rerun tests

**Pull request merged**